app/assets/javascripts/components/ChallengeModal/LockscreenWorkspaceSwitcher.tsx

```diff
@@ -0,0 +1,82 @@
+import { ApplicationGroup } from '@/ui_models/application_group';
+import { AppState } from '@/ui_models/app_state';
+import {
+  calculateSubmenuStyle,
+  SubmenuStyle,
+} from '@/utils/calculateSubmenuStyle';
+import { FunctionComponent } from 'preact';
+import { useEffect, useRef, useState } from 'preact/hooks';
+import { WorkspaceSwitcherMenu } from '../AccountMenu/WorkspaceSwitcher/WorkspaceSwitcherMenu';
+import { Button } from '../Button';
+import { Icon } from '../Icon';
+import { useCloseOnClickOutside } from '../utils';
+
+type Props = {
+  mainApplicationGroup: ApplicationGroup;
+  appState: AppState;
+};
+
+export const LockscreenWorkspaceSwitcher: FunctionComponent<Props> = ({
+  mainApplicationGroup,
+  appState,
+}) => {
+  const buttonRef = useRef<HTMLButtonElement>(null);
+  const menuRef = useRef<HTMLDivElement>(null);
+  const containerRef = useRef<HTMLDivElement>(null);
+  const [isOpen, setIsOpen] = useState(false);
+  const [menuStyle, setMenuStyle] = useState<SubmenuStyle>();
+
+  useCloseOnClickOutside(containerRef, () => setIsOpen(false));
+
+  const toggleMenu = () => {
+    if (!isOpen) {
+      const menuPosition = calculateSubmenuStyle(buttonRef.current);
+      if (menuPosition) {
+        setMenuStyle(menuPosition);
+      }
+    }
+
+    setIsOpen(!isOpen);
+  };
+
+  useEffect(() => {
+    if (isOpen) {
+      setTimeout(() => {
```

Comment:
Could we somehow avoid `setTimeout` here? I know doing that sometimes is very hard, but the fewer timeouts we have in our code the less headaches we'll have in the future 🙂